##  Import Libraries

In [123]:
# linear algebra
import numpy as np 

# data processing
import pandas as pd 

# data visualization
import seaborn as sns
%matplotlib inline
from matplotlib import pyplot as plt
from matplotlib import style


from sklearn.model_selection import train_test_split

from sklearn.ensemble import RandomForestClassifier

from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score,confusion_matrix

from sklearn.linear_model import LogisticRegression

from sklearn import neighbors
from sklearn.metrics import classification_report

## Load the data

In [124]:
addr1 = "train.csv"
addr2 ="destinations.csv"
addr3 ="test.csv"


train_df = pd.read_csv(addr1, nrows=200000)
destinations_df = pd.read_csv(addr2, nrows=100000)
test_df= pd.read_csv(addr3, nrows=200000)

In [125]:
train_df.shape

(200000, 24)

In [126]:
destinations_df.shape

(62106, 150)

In [127]:
train_df.describe()

site_name  posa_continent  user_location_country  \
count  200000.000000   200000.000000          200000.000000   
mean        8.902020        2.657040              85.249490   
std        11.909226        0.752855              54.458165   
min         2.000000        0.000000               0.000000   
25%         2.000000        3.000000              66.000000   
50%         2.000000        3.000000              66.000000   
75%        11.000000        3.000000              69.000000   
max        53.000000        4.000000             239.000000   

       user_location_region  user_location_city  orig_destination_distance  \
count         200000.000000       200000.000000              130363.000000   
mean             315.726450        27965.756295                1953.335631   
std              209.417337        16792.147285                2193.421104   
min                0.000000            3.000000                   0.005600   
25%              174.000000        13442.000000                 306.000350   
50%              314.000000        27655.000000                1123.419100   
75%              395.000000        42931.000000                2568.067150   
max             1025.000000        56498.000000               11641.224200   

             user_id      is_mobile     is_package       channel  ...  \
count  200000.000000  200000.000000  200000.000000  200000.00000  ...   
mean   380280.016485       0.136000       0.255325       5.81500  ...   
std    212464.047443       0.342789       0.436045       3.75299  ...   
min        12.000000       0.000000       0.000000       0.00000  ...   
25%    181983.000000       0.000000       0.000000       2.00000  ...   
50%    391007.000000       0.000000       0.000000       9.00000  ...   
75%    574353.000000       0.000000       1.000000       9.00000  ...   
max    740408.000000       1.000000       1.000000      10.00000  ...   

       srch_children_cnt    srch_rm_cnt  srch_destination_id  \
count      200000.000000  200000.000000        200000.000000   
mean            0.328000       1.116885         14591.808115   
std             0.723387       0.457574         11047.396255   
min             0.000000       0.000000             8.000000   
25%             0.000000       1.000000          8267.000000   
50%             0.000000       1.000000         11319.000000   
75%             0.000000       1.000000         19352.000000   
max             9.000000       8.000000         65035.000000   

       srch_destination_type_id     is_booking            cnt  \
count             200000.000000  200000.000000  200000.000000   
mean                   2.603865       0.083345       1.480630   
std                    2.165358       0.276404       1.202889   
min                    1.000000       0.000000       1.000000   
25%                    1.000000       0.000000       1.000000   
50%                    1.000000       0.000000       1.000000   
75%                    5.000000       0.000000       2.000000   
max                    9.000000       1.000000      59.000000   

       hotel_continent  hotel_country  hotel_market  hotel_cluster  
count    200000.000000  200000.000000  200000.00000  200000.000000  
mean          3.176690      80.292910     601.35258      49.637945  
std           1.628104      55.634115     509.62625      28.985590  
min           0.000000       0.000000       0.00000       0.000000  
25%           2.000000      50.000000     180.00000      25.000000  
50%           2.000000      50.000000     599.00000      49.000000  
75%           4.000000     105.000000     701.00000      73.000000  
max           6.000000     212.000000    2117.00000      99.000000  

[8 rows x 21 columns]

In [128]:
train_df.columns

Index(['date_time', 'site_name', 'posa_continent', 'user_location_country',
       'user_location_region', 'user_location_city',
       'orig_destination_distance', 'user_id', 'is_mobile', 'is_package',
       'channel', 'srch_ci', 'srch_co', 'srch_adults_cnt', 'srch_children_cnt',
       'srch_rm_cnt', 'srch_destination_id', 'srch_destination_type_id',
       'is_booking', 'cnt', 'hotel_continent', 'hotel_country', 'hotel_market',
       'hotel_cluster'],
      dtype='object')

In [129]:
train_df.isnull().sum()

date_time                        0
site_name                        0
posa_continent                   0
user_location_country            0
user_location_region             0
user_location_city               0
orig_destination_distance    69637
user_id                          0
is_mobile                        0
is_package                       0
channel                          0
srch_ci                        177
srch_co                        177
srch_adults_cnt                  0
srch_children_cnt                0
srch_rm_cnt                      0
srch_destination_id              0
srch_destination_type_id         0
is_booking                       0
cnt                              0
hotel_continent                  0
hotel_country                    0
hotel_market                     0
hotel_cluster                    0
dtype: int64

In [130]:
train_df.isna().sum()
train_df.isnull().any()
train_df = train_df.dropna()

In [139]:
train_df['srch_ci'] = pd.to_datetime(train_df['srch_ci'])
train_df['srch_co'] = pd.to_datetime(train_df['srch_co'])
train_df['date_time'] = pd.to_datetime(train_df['date_time'])
train_df['Cin_day'] = train_df["srch_ci"].apply(lambda x: x.day)
train_df['Cin_month'] = train_df["srch_ci"].apply(lambda x: x.month)
train_df['Cin_year'] = train_df["srch_ci"].apply(lambda x: x.year)

In [135]:
train_df.shape

(130238, 27)

In [133]:
Y = train_df['is_booking']
X = train_df.drop(['date_time','orig_destination_distance','srch_ci','srch_co','is_booking'],axis=1)
X.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 130238 entries, 0 to 199996
Data columns (total 22 columns):
 #   Column                    Non-Null Count   Dtype
---  ------                    --------------   -----
 0   site_name                 130238 non-null  int64
 1   posa_continent            130238 non-null  int64
 2   user_location_country     130238 non-null  int64
 3   user_location_region      130238 non-null  int64
 4   user_location_city        130238 non-null  int64
 5   user_id                   130238 non-null  int64
 6   is_mobile                 130238 non-null  int64
 7   is_package                130238 non-null  int64
 8   channel                   130238 non-null  int64
 9   srch_adults_cnt           130238 non-null  int64
 10  srch_children_cnt         130238 non-null  int64
 11  srch_rm_cnt               130238 non-null  int64
 12  srch_destination_id       130238 non-null  int64
 13  srch_destination_type_id  130238 non-null  int64
 14  cnt                 

###  Split the data into Train and Test ( 15% Test )

In [136]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.15, random_state=1)
print(len(X_train))
print(len(X_test))

110702
19536


###  RandomForestClassifier

In [119]:
clf = RandomForestClassifier(n_estimators=80)
clf.fit(X_train, Y_train)
y_pred=clf.predict(X_test)
acc=accuracy_score(Y_test,y_pred)


print("Accuracy:" ,acc)
print( "confusion_matrix")
print( confusion_matrix(Y_test, y_pred))

print(classification_report(Y_test, y_pred))
print(clf.feature_importances_)

Accuracy: 0.8642902475142796
confusion_matrix
[[8163  455]
 [ 828    8]]
              precision    recall  f1-score   support

           0       0.91      0.95      0.93      8618
           1       0.02      0.01      0.01       836

    accuracy                           0.86      9454
   macro avg       0.46      0.48      0.47      9454
weighted avg       0.83      0.86      0.85      9454

[0.01200692 0.00342927 0.01050118 0.06411019 0.09032888 0.09046352
 0.00936114 0.01060071 0.03503213 0.03074491 0.01858326 0.01091458
 0.09193735 0.02180172 0.04554937 0.01025126 0.02139675 0.08198315
 0.14808141 0.11218875 0.06165254 0.01908099]


###  LogisticRegression

In [120]:
lr = LogisticRegression()
lr.fit(X_train, Y_train)
y_predict = lr.predict(X_test)
acc=accuracy_score(Y_test,y_predict)

print("Accuracy:" ,acc)
print( "confusion_matrix")
print( confusion_matrix(Y_test, y_predict))

print(classification_report(Y_test, y_predict))

Accuracy: 0.9115718214512376
confusion_matrix
[[8618    0]
 [ 836    0]]
              precision    recall  f1-score   support

           0       0.91      1.00      0.95      8618
           1       0.00      0.00      0.00       836

    accuracy                           0.91      9454
   macro avg       0.46      0.50      0.48      9454
weighted avg       0.83      0.91      0.87      9454



C:\Users\ragun\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragun\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
C:\Users\ragun\anaconda3\lib\site-packages\sklearn\metrics\_classification.py:1245: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


###  KNeighborsClassifier

In [121]:
knn = neighbors.KNeighborsClassifier()
knn.fit(X_train,Y_train)
y_predict = knn.predict(X_test)
acc=accuracy_score(Y_test,y_predict)

print("Accuracy:" ,acc)
print( "confusion_matrix")
print( confusion_matrix(Y_test, y_predict))

print(classification_report(Y_test, y_predict))

Accuracy: 0.9028982441294691
confusion_matrix
[[8527   91]
 [ 827    9]]
              precision    recall  f1-score   support

           0       0.91      0.99      0.95      8618
           1       0.09      0.01      0.02       836

    accuracy                           0.90      9454
   macro avg       0.50      0.50      0.48      9454
weighted avg       0.84      0.90      0.87      9454



### Conclusion

1. Loaded the data

2. Performed the EDA and removed some NA values

3. Splited the data into Test and Training in the Ratio of 85% and 15%

4. Used Random Forest Model to verify and the accuracy is 91%

5. Used Logestic Regression Model to verify and the accuracy is 86%

5. Used KNeighborsClassifier Model to verify and the accuracy is 90%